pkg/controller/jobs/pod/pod_controller.go

```diff
@@ -723,47 +721,98 @@ func isPodRunnableOrSucceeded(p *corev1.Pod) bool {
+		if c.Type == corev1.ContainersReady {
+			if c.Status == corev1.ConditionFalse && c.Reason == string(corev1.PodFailed) {
+				lastTransition = c.LastTransitionTime
+			}
+			break
+		}
+	}
+	deletionTime := ptr.Deref(p.DeletionTimestamp, metav1.Now())
+	if lastTransition.Before(&deletionTime) {
+		return lastTransition.Time
+	}
+	return deletionTime.Time
+}
+
 // cleanupExcessPods will delete and finalize pods created last if the number of
 // activePods is greater than the totalCount value.
-func (p *Pod) cleanupExcessPods(ctx context.Context, c client.Client, r record.EventRecorder, totalCount int, activePods []corev1.Pod) error {
+func (p *Pod) cleanupExcessPods(ctx context.Context, c client.Client, r record.EventRecorder, totalCount int, activePods []corev1.Pod, inactivePods []corev1.Pod) error {
 	log := ctrl.LoggerFrom(ctx)
 
-	extraPodsCount := len(activePods) - totalCount
+	extraInactivePodsCount := min(len(inactivePods), len(inactivePods)+len(activePods)-totalCount)
+	extraActivePodsCount := len(activePods) - totalCount
 
-	if extraPodsCount <= 0 {
+	if extraActivePodsCount <= 0 && extraInactivePodsCount <= 0 {
 		return nil
 	}
 	// Do not clean up more pods until observing previous operations
 	if !p.satisfiedExcessPods {
 		return errPendingOps
 	}
 
-	// Sort active pods by creation timestamp
-	sort.Slice(activePods, func(i, j int) bool {
-		pi := &activePods[i]
-		pj := &activePods[j]
-		iFin := slices.Contains(pi.Finalizers, PodFinalizer)
-		jFin := slices.Contains(pj.Finalizers, PodFinalizer)
-		// Prefer to keep pods that have a finalizer.
-		if iFin != jFin {
-			return iFin
-		}
-		iGated := gateIndex(pi) != gateNotFound
-		jGated := gateIndex(pj) != gateNotFound
-		// Prefer to keep pods that aren't gated.
-		if iGated != jGated {
-			return !iGated
-		}
-		return pi.ObjectMeta.CreationTimestamp.Before(&pj.ObjectMeta.CreationTimestamp)
-	})
+	var extraPods []corev1.Pod
+
+	// Remove inactive pods that have a replacement
+	if extraInactivePodsCount > 0 {
+		// Sort inactive pods
+		sort.Slice(activePods, func(i, j int) bool {
+			pi := &activePods[i]
+			pj := &activePods[j]
+			iFin := slices.Contains(pi.Finalizers, PodFinalizer)
+			jFin := slices.Contains(pj.Finalizers, PodFinalizer)
+			// Prefer to keep pods that have a finalizer.
+			// Add them to the end of the list.
+			if iFin != jFin {
+				return jFin
+			}
 
-	// Extract all the latest created extra pods
-	extraPods := activePods[len(activePods)-extraPodsCount:]
-	extraPodsUIDs := utilslices.Map(extraPods, func(p *corev1.Pod) types.UID { return p.UID })
-	p.excessPodExpectations.ExpectUIDs(log, p.key, extraPodsUIDs)
+			iLastActive := lastActiveTime(pi)
+			jLastActive := lastActiveTime(pj)
 
-	// Finalize and delete the active pods created last
+			if iLastActive.Equal(jLastActive) {
+				return pi.CreationTimestamp.Before(&pj.CreationTimestamp)
+			}
+			return iLastActive.Before(jLastActive)
+		})
+
+		extraPods = inactivePods[:extraInactivePodsCount]
```

Comment:
Can you explore the option of not deleting these pods? Just removing the finalizer.

Then it would be up to the k8s garbage collector to take them.

This behavior would preserve the lifecycle of a Pod that users are accustomed to, avoiding potential surprises, or even allowing for some debugging of the Pod yaml before it's fully deleted.